cvmfs/catalog_mgr_client.cc

```diff
@@ -87,188 +88,281 @@ shash::Any ClientCatalogManager::GetRootHash() {
+    success_code = catalog::kLoadUp2Date;
+  } else if (local_newest_revision == 0 && mounted_catalogs_.size() > 0) {
+    // breadcrumb has no revision
+    // TODO(heretherebedragons) this branch can be removed in futur versions
+
+    // revisions are better, but if we dont have any we need to compare by
+    // timestamp (you can have multiple revisions in the same timestamp)
+    if (local_newest_timestamp < GetTimestampNoLock()) {
+      local_newest_hash = mounted_hash;
+      local_newest_revision = GetRevisionNoLock();
+      local_newest_timestamp = GetTimestampNoLock();
+      result->SetRootCtlgLocation(kCtlgLocationMounted);
+      success_code = catalog::kLoadUp2Date;
+    }
   }
 
-  // Load and verify remote checksum
+  // 3) Get remote root catalog (fails if remote catalog is older)
   manifest::Failures manifest_failure;
-  CachedManifestEnsemble ensemble(fetcher_->cache_mgr(), this);
-  manifest_failure = manifest::Fetch("", repo_name_, cache_last_modified,
-                                     &cache_hash, signature_mgr_,
+  UniquePtr<CachedManifestEnsemble> ensemble(
+                       new CachedManifestEnsemble(fetcher_->cache_mgr(), this));
+  manifest_failure = manifest::Fetch("", repo_name_, local_newest_timestamp,
+                                     &local_newest_hash, signature_mgr_,
                                      fetcher_->download_mgr(),
-                                     &ensemble);
-  if (manifest_failure != manifest::kFailOk) {
-    LogCvmfs(kLogCache, kLogDebug, "failed to fetch manifest (%d - %s)",
-             manifest_failure, manifest::Code2Ascii(manifest_failure));
-
-    LoadError success_code = catalog::kLoadUp2Date;
-
-    // Network unavailable but cached copy updated externally?
-    std::map<PathString, shash::Any>::const_iterator iter =
-      mounted_catalogs_.find(mountpoint);
-    if (iter != mounted_catalogs_.end()) {
-      if (breadcrumb.IsValid() && (iter->second != cache_hash)) {
-        success_code = catalog::kLoadNew;
+                                     ensemble.weak_ref());
+
+  if (manifest_failure == manifest::kFailOk) {
+    // server has newest revision or no valid local revision
+    if (ensemble->manifest->revision() > local_newest_revision
+          || local_newest_revision == manifest::Breadcrumb::kInvalidRevision
+          // if revision is 0 both local and server, load catalog from server
+          // as local is most likely just "initialized" without valid value
+          || (ensemble->manifest->revision() == 0 &&
+              local_newest_revision == 0)) {
+      result->SetHash(ensemble->manifest->catalog_hash());
+      result->SetRootCtlgRevision(ensemble->manifest->revision());
+      result->SetRootCtlgLocation(kCtlgLocationServer);
+      if (ensemble->manifest->has_alt_catalog_path()) {
+        fixed_alt_root_catalog_ = true;
+      } else {
+        fixed_alt_root_catalog_ = false;
       }
```

Comment:
Can be written in fewer lines as 

    fixed_alt_root_catalog_ = ensemble->manifest->has_alt_catalog_path();

Reply:
oops  not sure what i thought there